### PR TITLE
BUG: fix a regression in astropy 6.0.1 where using a dimensionless Quantity as an exponent could raise an exception

### LINKED ISSUE #16260: astropy 6.0.1 breaks using quantities as unit exponents
### Description

This works in astropy 6.0.0:

```
u.m**u.Quantity(2)
```

but raises an exception in astropy 6.0.1:

```
u.m**u.Quantity(2)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[2], line 1
----> 1 u.m**u.Quantity(2)

File ~/.local/conda/envs/gammapy-dev/lib/python3.11/site-packages/astropy/units/core.py:831, in UnitBase.__pow__(self, p)
    829 def __pow__(self, p):
    830     p = validate_power(p)
--> 831     return CompositeUnit(1, [self], [p], _error_check=False)

File ~/.local/conda/envs/gammapy-dev/lib/python3.11/site-packages/astropy/units/core.py:2341, in CompositeUnit.__init__(self, scale, bases, powers, decompose, decompose_bases, _error_check)
   2339         scale *= unit.scale**power
   2340         self._bases = unit.bases
-> 2341         self._powers = [
   2342             sanitize_power(operator.mul(*resolve_fractions(p, power)))
   2343             for p in unit.powers
   2344         ]
   2346     self._scale = sanitize_scale(scale)
   2347 else:
   2348     # Regular case: use inputs as preliminary scale, bases, and powers,
   2349     # then "expand and gather" identical bases, sanitize the scale, &c.

File ~/.local/conda/envs/gammapy-dev/lib/python3.11/site-packages/astropy/units/core.py:2342, in <listcomp>(.0)
   2339         scale *= unit.scale**power
   2340         self._bases = unit.bases
   2341         self._powers = [
-> 2342             sanitize_power(operator.mul(*resolve_fractions(p, power)))
   2343             for p in unit.powers
   2344         ]
   2346     self._scale = sanitize_scale(scale)
   2347 else:
   2348     # Regular case: use inputs as preliminary scale, bases, and powers,
   2349     # then "expand and gather" identical bases, sanitize the scale, &c.

File ~/.local/conda/envs/gammapy-dev/lib/python3.11/site-packages/astropy/units/utils.py:291, in sanitize_power(p)
    288 denom = getattr(p, "denominator", None)
    289 if denom is None:
    290     # This returns either a (simple) Fraction or the same float.
--> 291     p = maybe_simple_fraction(p)
    292     # If still a float, nothing more to be done.
    293     if isinstance(p, float):

File ~/.local/conda/envs/gammapy-dev/lib/python3.11/site-packages/astropy/units/utils.py:217, in maybe_simple_fraction(p, max_denominator)
    215 if p == 0 or p.__class__ is int or p.__class__ is Fraction:
    216     return p
--> 217 n, d = p.as_integer_ratio()
    218 a = n // d
    219 # Normally, start with 0,1 and 1,0; here we have applied first iteration.

File ~/.local/conda/envs/gammapy-dev/lib/python3.11/site-packages/astropy/units/quantity.py:1108, in Quantity.__getattr__(self, attr)
   1103 """
   1104 Quantities are able to directly convert to other units that
   1105 have the same physical type.
   1106 """
   1107 if not self._include_easy_conversion_members:
-> 1108     raise AttributeError(
   1109         f"'{self.__class__.__name__}' object has no '{attr}' member"
   1110     )
   1112 def get_virtual_unit_attribute():
   1113     registry = get_current_unit_registry().registry

AttributeError: 'Quantity' object has no 'as_integer_ratio' member
```

### Expected behavior

Bugfix releases don't introduce breaking changes

### How to Reproduce

```python
import astropy.units as u

u.m**u.Quantity(2)
```

### Versions

astropy 6.0.1

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #16260

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/units/tests/test_quantity.py [hunk 1]
```diff
@@ -468,6 +468,11 @@ def test_power(self):
         assert_array_almost_equal(new_quantity.value, 1489.355288, decimal=7)
         assert new_quantity.unit == u.Unit("m^3")
 
+    def test_quantity_as_power(self):
+        # raise unit to a dimensionless Quantity power
+        # regression test for https://github.com/astropy/astropy/issues/16260
+        assert u.m ** u.Quantity(2) == u.m**2
```
Comment (maxnoe) [inline astropy/units/tests/test_quantity.py:474]: Another case where this fails is with numpy integers:

```
In [4]: maybe_simple_fraction(np.int64(5))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 maybe_simple_fraction(np.int64(5))

File ~/Projects/astropy/astropy/units/utils.py:217, in maybe_simple_fraction(p, max_denominator)
    215 if p == 0 or p.__class__ is int or p.__class__ is Fraction:
    216     return p
--> 217 n, d = p.as_integer_ratio()
    218 a = n // d
    219 # Normally, start with 0,1 and 1,0; here we have applied first iteration.

AttributeError: 'numpy.int64' object has no attribute 'as_integer_ratio'
```

Comment (maxnoe) [reply astropy/units/tests/test_quantity.py:474]:
```
@pytest.mark.parametrize("exponent", [u.Quantity(2), np.uint64(2), np.int32(2), 2.0, np.float32(2.0)])
def test_unit_powers(exponent):
    assert u.m**exponent = u.m**2
```

File: astropy/units/utils.py [hunk 2]
```diff
@@ -214,7 +214,7 @@ def maybe_simple_fraction(p, max_denominator=100):
     """
     if p == 0 or p.__class__ is int or p.__class__ is Fraction:
         return p
-    n, d = p.as_integer_ratio()
```
Comment (maxnoe) [inline astropy/units/utils.py:217]: Comment to the if above, which is not new code:

the python data model API for "integer-like" things is `__index__`, so you could check for:

```
if hasattr(p, "__index__"):
    return p.__index__()
```

https://docs.python.org/3/reference/datamodel.html#object.__index__

Comment (maxnoe) [reply astropy/units/utils.py:217]: that e.g. works for the numpy integer types which currently fail. This is of course fixed with your addition of `float(p)`, but then you run the algorithm needlessly for integers.

File: astropy/units/tests/test_quantity.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/units/tests/test_quantity.py]: good point, I've merged my test with the previous one and parametrized it to better cover regressions

File: astropy/units/utils.py (hunk 2 above)
Comment (neutrinoceros) [reply astropy/units/utils.py:217]: I think there may be two different concerns here:
- fixing a regression
- making sure performances are optimal

In my opinion it's best to keep this PR focused on the first goal, but we can certainly tackle the second one in a follow up PR !
In the meantime, if anyone is worried about performance regressions from my proposed patch, we could run benchmarks on this PR

File: astropy/units/tests/test_quantity.py [hunk 3]
```diff
@@ -458,13 +458,18 @@ def test_commutativity(self):
         assert self.q1 / u.s == u.Quantity(11.42, "m/s")
         assert u.s / self.q1 == u.Quantity(1 / 11.42, "s/m")
 
-    def test_power(self):
+    @pytest.mark.parametrize(
+        "exponent_type", [float, np.uint64, np.int32, np.float32, u.Quantity]
+    )
+    def test_power(self, exponent_type):
```
Comment (maxnoe) [inline astropy/units/tests/test_quantity.py:464]: This test does not actually test the regression, i.e. it doesn't fail on 6.0.1.

The difference is taking a power of a quantity vs. a power of a unit:

```
q = 5 * u.m
q = q**u.Quantity(3)
```
work in astropy 6.0.1,

```
q = 125 * u.m**u.Quantity(3)
```
does not
```

**Review by maxnoe — CHANGES_REQUESTED** [review]

File: astropy/units/tests/test_quantity.py (hunk 3 above)
Comment (neutrinoceros) [reply astropy/units/tests/test_quantity.py:464]: Ah. My first test did correctly fail on 6.0.1
I will rehash this after lunch, thanks a lot for your attention to details !

Comment (neutrinoceros) [reply astropy/units/tests/test_quantity.py]: done !
note that maybe_simple_fraction is (I think) private API, so I focused my test on just public API by exercising various types of exponents in m**2, but the only case that actually failed on main (or v6.0.x) is still m**Quantity(2)

Comment (neutrinoceros) [reply astropy/units/tests/test_quantity.py]: Actually, we got the same regression with `u.Masked(2)`, so I included it in the test too

File: astropy/units/utils.py (hunk 2 above)
Comment (pllim) [reply astropy/units/utils.py:217]: I added `benchmark`  label. FYI.

**Review by mhvk — APPROVED** [review]
Thanks, that fix looks good. Since we already have a quick path for plain ints, and the hit for float (likely the next most common one) is only 10 ns or so, I think this is even about as good as one can hope for performance-wise.

**Comment by neutrinoceros** [comment]
this was triaged to 6.0.2 bug the backport label wasn't added, so I'll assume it was a mistake and I'll open a manual backport PR

**Comment by pllim** [comment]
FWIW

```
| Change   | Before [495efdb3]    | After [7df147cb]    |   Ratio | Benchmark (Parameter)                                      |
|----------|----------------------|---------------------|---------|------------------------------------------------------------|
| -        | 300±6μs              | 227±4μs             |    0.76 | units.TimeQuantityOpLargeArraySameUnit.time_quantity_equal |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

**Pull request merged**